EKF/airspeed_fusion.cpp

```diff
@@ -46,152 +47,137 @@
 
-		// Calculate measurement innovation
-		_airspeed_innov = v_tas_pred -
-				  _airspeed_sample_delayed.true_airspeed;
+		return;
+	}
 
-		// Compute the ratio of innovation to gate size
-		_tas_test_ratio = sq(_airspeed_innov) / (sq(fmaxf(_params.tas_innov_gate, 1.0f)) * _airspeed_innov_var);
+	// Observation Jacobians
+	Hfusion.at<4>() = HK4;
+	Hfusion.at<5>() = HK5;
+	Hfusion.at<6>() = HK3*vd;
+	Hfusion.at<22>() = -HK4;
+	Hfusion.at<23>() = -HK5;
+
+	if (!update_wind_only) {
+		// we have no other source of aiding, so use airspeed measurements to correct states
+		for (unsigned row = 0; row <= 3; row++) {
+			Kfusion(row) = HK16*(HK0*P(4,row) - HK0*P(row,22) + HK1*P(5,row) - HK1*P(row,23) + P(6,row)*vd);
+		}
 
-		// If the innovation consistency check fails then don't fuse the sample and indicate bad airspeed health
-		if (_tas_test_ratio > 1.0f) {
-			_innov_check_fail_status.flags.reject_airspeed = true;
-			return;
+		Kfusion(4) = HK12*HK16;
+		Kfusion(5) = HK16*HK9;
+		Kfusion(6) = HK16*HK7;
 
-		} else {
-			_innov_check_fail_status.flags.reject_airspeed = false;
+		for (unsigned row = 7; row <= 21; row++) {
+			Kfusion(row) = HK16*(HK0*P(4,row) - HK0*P(row,22) + HK1*P(5,row) - HK1*P(row,23) + P(6,row)*vd);
 		}
+	}
 
-		// apply covariance correction via P_new = (I -K*H)*P
-		// first calculate expression for KHP
-		// then calculate P - KHP
-		SquareMatrix24f KHP;
-		float KH[5];
-
-		for (unsigned row = 0; row < _k_num_states; row++) {
-
-			KH[0] = Kfusion(row) * H_TAS(4);
-			KH[1] = Kfusion(row) * H_TAS(5);
-			KH[2] = Kfusion(row) * H_TAS(6);
-			KH[3] = Kfusion(row) * H_TAS(22);
-			KH[4] = Kfusion(row) * H_TAS(23);
-
-			for (unsigned column = 0; column < _k_num_states; column++) {
-				float tmp = KH[0] * P(4,column);
-				tmp += KH[1] * P(5,column);
-				tmp += KH[2] * P(6,column);
-				tmp += KH[3] * P(22,column);
-				tmp += KH[4] * P(23,column);
-				KHP(row,column) = tmp;
-			}
-		}
+	Kfusion(22) = HK15*HK16;
+	Kfusion(23) = HK14*HK16;
 
-		const bool healthy = checkAndFixCovarianceUpdate(KHP);
+	// Calculate measurement innovation
+	_airspeed_innov = v_tas_pred - _airspeed_sample_delayed.true_airspeed;
 
-		_fault_status.flags.bad_airspeed = !healthy;
+	// Compute the ratio of innovation to gate size
+	_tas_test_ratio = sq(_airspeed_innov) / (sq(fmaxf(_params.tas_innov_gate, 1.0f)) * _airspeed_innov_var);
 
-		if (healthy) {
-			// apply the covariance corrections
-			P -= KHP;
+	// If the innovation consistency check fails then don't fuse the sample and indicate bad airspeed health
+	if (_tas_test_ratio > 1.0f) {
+		_innov_check_fail_status.flags.reject_airspeed = true;
+		return;
 
-			fixCovarianceErrors(true);
+	} else {
+		_innov_check_fail_status.flags.reject_airspeed = false;
+	}
 
-			// apply the state corrections
-			fuse(Kfusion, _airspeed_innov);
+	// Airspeed measurement sample has passed check so record it
+	_time_last_arsp_fuse = _time_last_imu;
```

Comment:
The  `_time_last_arsp_fuse ` is also set below. I argue that we should only update it once it passes also the covariance health checks.